### PR TITLE
clang-tidy clang-tidy-modernize-use-emplace

### PR DESCRIPTION
**Pull request by JohnCoconut**
A example of clang-tidy fix is shown below.

Before:

```cpp
outPoints.push_back(Vec2(pt.X/PRECISION, pt.Y/PRECISION));
```

After

```cpp
outPoints.emplace_back(pt.X/PRECISION, pt.Y/PRECISION);
```

Reference: https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-emplace.html

### CODE REVIEW

File: cocos/editor-support/cocostudio/CCColliderDetector.cpp [hunk 1]
```diff
@@ -196,7 +196,7 @@ void ColliderDetector::addContourData(ContourData *contourData)
     unsigned long num = contourData->vertexList.size();
     for (unsigned long i = 0; i < num; i++)
     {
-        calculatedVertexList.push_back(Vec2());
+        calculatedVertexList.emplace_back();
```
Comment (minggo) [inline cocos/editor-support/cocostudio/CCColliderDetector.cpp:200]: May be just use reserve here?

Comment (JohnCoconut) [reply cocos/editor-support/cocostudio/CCColliderDetector.cpp:200]: Yes. `reserve` can be used here. However, upon reading the context around this line, I found it hard to understand.

`calculatedVertexList` is a reference to a member,

```cpp
std::vector<Vec2> &calculatedVertexList = colliderBody->_calculatedVertexList;
```

It's initialized in

```cpp
#if ENABLE_PHYSICS_SAVE_CALCULATED_VERTEX
    _calculatedVertexList = Array::create();
    CC_SAFE_RETAIN(_calculatedVertexList);
#endif
```

Does the code in the macro ever run? I am not familiar with the `ColliderBody` class. The following line wouldn't compile.

```cpp
std::vector<cocos2d::Vec2> vertex_list = Array::create();
```

Comment (minggo) [reply cocos/editor-support/cocostudio/CCColliderDetector.cpp:200]: I am not familiar with it too. But the codes is `calculatedVertexList.emplace_back();` is not easy to understand. It seems it emplace back nothing.

Comment (JohnCoconut) [reply cocos/editor-support/cocostudio/CCColliderDetector.cpp:200]: `emplace_back()` means default constructor being used. For example,

```cpp
std::vector<std::string> v;
v.emplace_back();
```
appends a empty string to v.

Comment (minggo) [reply cocos/editor-support/cocostudio/CCColliderDetector.cpp:200]: Yep, i understand.

**Comment by minggo** [comment]
@JohnCoconut will you do the same thing in v4 too? Your PRs are very good to v4 too.

**Comment by JohnCoconut** [comment]
Do you want to create another clang-tidy tests build in v4 branch?

**Comment by JohnCoconut** [comment]
What's the status of v4 branch? I checked out v4 branch, and it didn't compile with the following error.

```c++
../cocos/scripting/lua-bindings/auto/lua_cocos2dx_backend_auto.cpp:2014:35: error: no matching function for call to ‘cocos2d::backend::VertexLayout::setLayout(unsigned int&, cocos2d::backend::VertexStepMode&)’
 2014 |         cobj->setLayout(arg0, arg1);
      |                                   ^
In file included from ../cocos/scripting/lua-bindings/auto/lua_cocos2dx_backend_auto.cpp:5:
../cocos/renderer/backend/VertexLayout.h:82:10: note: candidate: ‘void cocos2d::backend::VertexLayout::setLayout(unsigned int)’
   82 |     void setLayout(unsigned int stride);
      |          ^~~~~~~~~
../cocos/renderer/backend/VertexLayout.h:82:10: note:   candidate expects 1 argument, 2 provided
```

**Comment by minggo** [comment]
> Do you want to create another clang-tidy tests build in v4 branch?

What did you mean another clang-tidy tests?

About the error, i will check it.

**Comment by JohnCoconut** [comment]
> What did you mean another clang-tidy tests?

clang-tidy tests isn't in v4 branch. It's only available in v3 branch. 

https://github.com/cocos2d/cocos2d-x/blob/21bf81848244c758438ce48d568d3acbe0812d43/tools/travis-scripts/run-script.sh#L40-L50

**Comment by minggo** [comment]
Yep, we wan to add it in v4 too.

**Pull request merged**